### ISSUE TITLE
Amazon Yes Intent with Alexa (Lambda nodejs)

### ISSUE THREAD

**Issue by amfaraj94** [alexa/alexa-skills-kit-sdk-for-nodejs#343]
Hey guys.

We're working on this school project for Alexa and I would really appreciate if someone can help me out with this. So here is the section I want to work on in my code

```
var callAPEXDistMethod = function (callbackDistMethod) {

    var url = "https://apex.oracle.com/pls/apex/kelloggwmu/kellogg.cnhl4iqabzv5.us-east-1.rds.amazonaws.comDistMethod/";
    var req = https.get(url, (res) => {
        var body = "";
        res.on("data", (chunk) => {
            body += chunk;
        });
        res.on("end",  () => {
            var result = JSON.parse(body);

            callbackDistMethod({"speech":result.items[0].message + ". Would you like to know what was the most common distribution method in 2016?" });

            //callback('test');
        });
    }).on("error", (error) => {
        //callback(err);
        console.log('error');
    });
};

```

So basically what this code does, is that it calls in an API to get a response.
If you look at this part:

 callbackDistMethod({"speech":result.items[0].message + ". Would you like to know what was the most common distribution method in 2016?" });

The code gets the message from the API URL link on top.
Alexa gives us a response from the API, and then we have a following up question.

(". Would you like to know what was the most common distribution method in 2016?" )

I would like to use the YesIntent (or anything that works) somehow and make it so that if the user says "Yes", it would call in a different API to get the response for that question.

I'm really not sure how to do this and I'll greatly appreciate if someone can help me out.

**Comment by McGern** [comment]
How does the api you are using get what year it? i.e. does the current distribution default to last year (2017), then you want it to ask for the previous year (2016)?

Also how are the rest of your intent handlers structured and how is the conversation flow structured?

You may want to re-jig your ajax request so it takes a year as well, and maybe also consider using promises instead of callbacks (or async, I haven't used it because Lambda was only on version 6 of node until recently, and I haven't had a chance to play yet).

If your skill is quite simple (i.e. there is only one purpose of the "yes intent", no other context), then you could do it all within the same "handlers" group rather than using dialogs.

It might be worth having a look at the "history buff" cookbook example, it is a bit more complex than what you want (maybe?), but has examples of ajax callbacks, filling slots and  pagination "next" intents (which might be similar to your "yes" intent).

https://github.com/alexa/alexa-cookbook/tree/master/handling-responses/progressive-response

If you are still stuck, give as much info as you are able and I will see if I can help at all with code examples.

**Comment by amfaraj94** [comment]
McGern thank you for your response.
You can take a look at the entire code in this link and hopefully you'll have a better idea:

https://github.com/amfaraj94/alexa-cookbook/blob/d9ea5fc7be9f77dbf5006e36f50072b178751c92/handling-responses/progressive-response/AlexaOracleAPI

For the API, we're using Oracle server on apex.oracle.com. We have a database there and we made queries that get the results from the database, and we have put the URL links for the query results on the code to make it generate the response. (Look at the URL at line 121 in the link above) Hopefully this makes sense. 

I have an intent called DistMethod, you can use ctrl + find DistMethod in the code to see what I did with this intent to make it generate a response from the API.

The entire code is working fine, I just want a new intent with the same concept as DistMethod to be executed after line 133 that answers this question that:
(". Would you like to know what was the most common distribution method in 2016?")

So basically this is what we want. The user will say "What is the most common distribution method for coupons?"to execute DistMethod intent.

After Alexa gives the user the response from the API, Alexa will say:
"(. Would you like to know what was the most common distribution method in 2016?")"

If the user says "Yes", we want a new intent to be executed with the same concept.
Only it will have a different URL link for the API.

I'm really not that familiar with Alexa I'm not sure how to use handlers in this code.
My colleague was able to make a separate skill with handlers, but linking it with the API section in the code is the hard part.

Thank you again for your interest in helping.
Please get back to me as soon as you can.

**Comment by McGern** [comment]
Sorry, just a short response for now. Got to get on with my work, will try a more detailed response soon.

It looks like you are going down the hard path and not used the alexa node sdk at all (or perhaps a very old one)? The file "AlexaOracleApi" has the api calls mixed in with the intent functions.

If you follow the pattern presented in this file
https://github.com/amfaraj94/alexa-cookbook/blob/d9ea5fc7be9f77dbf5006e36f50072b178751c92/handling-responses/progressive-response/lambda/custom/index.js
you will see that the intents are for taking slot inputs (if applicable), and preparing the output speech with some logic for slots etc, but they call the api and get a result which is then used in the intent. This makes for reusable methods and more readable code. Also the SDK takes out a lot of the hard work for intent matching.

If you want to check it out, it is worth trying out the ASK CLI, which (once it's setup) makes it reasonably easy to create a new project that has a base template and then to link the Alexa Skill to the lambda function and deploy it all to AWS.

I usually start with something basic, make sure the intents work with the code, then add to it gradually. I also used the node package lamdba-local https://www.npmjs.com/package/lambda-local to test locally, which takes out some of the headache of having to keep uploading to AWS.

**BUT**, if you want a quick-and-dirty fix, you could try this

1. Add a "Yes" intent from the list of builtin intents in the Alexa skill builder. You should then get an intent "pointer" to an AMAZON.YesIntent handler
2. In the "onIntent" function add something to handle the AMAZON.YesIntent to line 84(?)

```
// dispatch custom intents to handlers here
if (intentName == 'DistMethod') {
    handleChartRequestDistMethod(intent, session, callback);
}
else if (intentName == 'NumberOfRecords') {
    handleChartRequestNumberOfRecords(intent, session, callback);
}
else if (intentName == 'AMAZON.YesIntent') {
    handleChartRequestDistMethod2016(intent,session,callback);
}

```
3. Create the function that will handle the intent call       


This should get you over the line, but isn't particularly useful because your conversation is quite hardcoded. Updating your skill to be more dynamic will also be a chore because the api and intent handling is mixed together. I would suggest seeing if you can get this working first, then start again using a new project and applying your existing code/logic. It should take too long, because most of what you have written is handled by the Alexa SDK.

**Comment by McGern** [comment]
How did you get on?

**Comment by benedekh** [comment]
@amfaraj94  just a hint: probably you could implement your skill following the state machine pattern. You could introduce some states and so define a context in which the response is comprehended. In this way you can say the answer for the question can be a Yes/No/something else, and this answer will be interpreted in that state.

As an example, have a look at the [Skill Game](https://github.com/alexa/skill-sample-nodejs-quiz-game/blob/master/lambda/custom/index.js#L237) project.
